### PR TITLE
feat: generate Twilio.ts file

### PR DESCRIPTION
**Pull request by beebzz**
Relates to https://code.hq.twilio.com/twilio/librarian/pull/381

### CODE REVIEW

**Review by claudiachua — APPROVED** [review]

File: lib/rest/Twilio.ts [hunk 1]
```diff
@@ -0,0 +1,934 @@
+'use strict'
```
Comment (claudiachua) [inline lib/rest/Twilio.ts:1]: Do we have a header to indicate that this is generated by openapi spec, similar to "this code is generated by yoyodyne"

**Review by claudiachua — APPROVED** [review]

File: lib/rest/Twilio.ts [hunk 2]
```diff
@@ -0,0 +1,934 @@
...
+ * @param {string} password -
+ *          The password used for authentication. This is normally auth token, but if using key/secret auth will be
+ *          the secret.
+ * @param {opts} [opts] - The options argument
+ *
+ * @returns {BaseTwilio} A new instance of BaseTwilio
+ */
+/* jshint ignore:end */
+
+class BaseTwilio {
```
Comment (childish-sambino) [inline lib/rest/Twilio.ts:54]: Should move this into a separate file so we all this hand-crafted code doesn't need to go in the template.

File: lib/rest/Twilio.ts [hunk 3]
```diff
@@ -0,0 +1,934 @@
...
+        this._proxyv1
+        this._ipmessagingv2
+        this._pricingv1
+    }
+
+  }
+}
+
+Object.defineProperty(Twilio.prototype,
+'numbersv2', {
```
Comment (childish-sambino) [inline lib/rest/Twilio.ts:411]: You should be able to define these from within the class definition itself: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get

File: lib/rest/Twilio.ts (hunk 1 above)
Comment (beebzz) [reply lib/rest/Twilio.ts:1]: We don't currently do that for OAI-generated files, but would be a good standup/eng sync question

File: lib/base/BaseTwilio.ts [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+import RequestClient from '../base/RequestClient'; /* jshint ignore:line */
+import { HttpMethod } from '../interfaces'; /* jshint ignore:line */
+
+var os = require('os');  /* jshint ignore:line */
+var url = require('url');  /* jshint ignore:line */
+var moduleInfo = require('../../package.json');  /* jshint ignore:line */
+var util = require('util');  /* jshint ignore:line */
+var RestException = require('../base/RestException');  /* jshint ignore:line */
+
+export interface opts {
```
Comment (childish-sambino) [inline lib/base/BaseTwilio.ts:12]: Think this name should be more specific.

File: lib/base/BaseTwilio.ts [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    accountSid?: string
+    env?: NodeJS.ProcessEnv
+    edge?: string
+    region?: string
+    lazyLoading?: boolean
+    logLevel?: string
+    userAgentExtensions?: string[]
+}
+
+interface requestOpts {
```
Comment (childish-sambino) [inline lib/base/BaseTwilio.ts:23]: Probably should export this.

File: lib/rest/api/v2010/account/recording.d.ts [hunk 6]
```diff
@@ -255,6 +255,7 @@ interface RecordingResource {
   duration: string;
   encryption_details: object;
   error_code: number;
+  media_url: string;
```
Comment (childish-sambino) [inline lib/rest/api/v2010/account/recording.d.ts:258]: Shouldn't be pulling in these changes in the PR.

File: lib/rest/Twilio.ts [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+  _wirelessv1: any
+  _proxyv1: any
+  _ipmessagingv2: any
+  _pricingv1: any
+
+  constructor(username: string, password: string, opts: opts) {
+    super(username, password, opts);
+
+    //Domains
+    this._numbersv2 = undefined
```
Comment (childish-sambino) [inline lib/rest/Twilio.ts:124]: Don't think this is needed. Should already be `undefined`.

File: lib/rest/Twilio.ts [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+    this._lookupsv1 = undefined
+    this._autopilotv1 = undefined
+    this._taskrouterv1 = undefined
+    this._wirelessv1 = undefined
+    this._proxyv1 = undefined
+    this._ipmessagingv2 = undefined
+    this._pricingv1 = undefined
+
+    if (!opts.lazyLoading) {
+        this._numbersv2
```
Comment (childish-sambino) [inline lib/rest/Twilio.ts:162]: These should be calling the non-underscored property.

File: lib/rest/Twilio.ts [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+        this._chatv1
+        this._conversationsv1
+        this._bulkexportsv1
```
Comment (childish-sambino) [inline lib/rest/Twilio.ts:165]: Should see about sorting these. Could just need a `sorted()` wrapper on the list return from `listdir`. See the recent Terraform provider changes in librarian for reference.

File: lib/base/BaseTwilio.ts [hunk 10]
```diff
@@ -0,0 +1,228 @@
+'use strict'
+
+import RequestClient from '../base/RequestClient'; /* jshint ignore:line */
+import { HttpMethod } from '../interfaces'; /* jshint ignore:line */
+
+var os = require('os');  /* jshint ignore:line */
```
Comment (childish-sambino) [inline lib/base/BaseTwilio.ts:6]: nit: should use `const` over `var` for most cases.

File: lib/rest/Twilio.ts [hunk 11]
```diff
@@ -0,0 +1,481 @@
...
+ *
+ * @returns {Twilio} A new instance of Twilio client
+ */
+/* jshint ignore:end */
+
+
+export class Twilio extends BaseTwilio {
+  username: string
+  password: string
+  opts: clientOpts
```
Comment (childish-sambino) [inline lib/rest/Twilio.ts:82]: These are defined in the base class.

File: lib/base/RequestClient.d.ts [hunk 12]
```diff
@@ -52,6 +52,10 @@ declare namespace RequestClient {
      * Set to true to use the forever-agent
      */
     forever?: boolean;
+    /**
+     * Set to 'debug' to enable debug logging
+     */
+    logLevel?: string;
```
[the remark this answers is not in the document]
Comment (beebzz) [reply lib/base/RequestClient.d.ts:58]: When adding the log level to the request as is (without including it in the `RequestClient` datatype), this error pops up: `Object literal may only specify known properties, and 'logLevel' does not exist in type 'RequestOptions<object, object>'`

File: lib/rest/Twilio.ts [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+  _trusthubv1: any
+  _verifyv2: any
+  _videov1: any
+  _voicev1: any
+  _wirelessv1: any
+
+  constructor(username: string, password: string, opts: ClientOpts) {
+    super(username, password, opts);
+
+    if (!opts.lazyLoading) {
```
Comment (childish-sambino) [inline lib/rest/Twilio.ts:120]: Should be `this.opts.lazyLoading === false`

**Review by childish-sambino — APPROVED** [review]

**Review by childish-sambino — APPROVED** [review]

**Pull request merged**